File: src/main/java/net/artux/pdanetwork/service/user/UserServiceImpl.java
package net.artux.pdanetwork.service.user;

import lombok.RequiredArgsConstructor;
import net.artux.pdanetwork.entity.user.UserEntity;
import net.artux.pdanetwork.models.Status;
import net.artux.pdanetwork.models.security.SecurityUser;
import net.artux.pdanetwork.models.user.UserMapper;
import net.artux.pdanetwork.models.user.dto.AdminEditUserDto;
import net.artux.pdanetwork.models.user.dto.AdminUserDto;
import net.artux.pdanetwork.models.user.dto.RegisterUserDto;
import net.artux.pdanetwork.models.user.dto.UserDto;
import net.artux.pdanetwork.models.user.enums.Role;
import net.artux.pdanetwork.repository.user.UserRepository;
import net.artux.pdanetwork.service.email.EmailService;
import net.artux.pdanetwork.utills.RandomString;
import net.artux.pdanetwork.utills.security.AdminAccess;
import org.apache.commons.io.output.ByteArrayOutputStream;
import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.CellStyle;
import org.apache.poi.ss.usermodel.Row;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.core.env.Environment;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.time.Instant;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;
import java.util.UUID;

@Service
@Transactional
@RequiredArgsConstructor
public class UserServiceImpl implements UserService {

    private final Logger logger = LoggerFactory.getLogger(UserServiceImpl.class);

    private final UserRepository userRepository;
    private final EmailService emailService;
    private final UserValidator userValidator;
    private final PasswordEncoder passwordEncoder;
    private final UserMapper userMapper;

    private final Map<String, RegisterUserDto> registerUserMap = new HashMap<>();
    private final Timer timer = new Timer();
    private final Environment environment;
    private final RandomString randomString = new RandomString();

    private boolean emailConfirmation = false;

    @Override
    public Status registerUser(RegisterUserDto newUser) {
        Status status = userValidator.checkUser(newUser);
        if (status.isSuccess()) {
            if (!registerUserMap.containsValue(newUser))
                try {
                    String token = generateToken(newUser);
                    if (emailConfirmation) {
                        emailService.sendConfirmLetter(newUser, token);
                        status = new Status(true, "Проверьте почту.");
                    }else {
                        handleConfirmation(token);
                        status = new Status(true, "Учетная запись зарегистрирована.");
                    }

                } catch (Exception e) {
                    logger.error("Registration", e);
                    status = new Status(false, "Не удалось отправить письмо на " + newUser.getEmail());
                }
            else
                status = new Status(false, "Пользователь ожидает регистрации, проверьте почту.");
        }

        return status;
    }

    private String generateToken(RegisterUserDto user) {
        String token = randomString.nextString();
        logger.info("Пользователь {} добавлен в лист ожидания регистрации с токеном {}, токен возможно использовать через сваггер.", user.getEmail(), token);
        registerUserMap.put(token, user);
        timer.schedule(new TimerTask() {
            @Override
            public void run() {
                registerUserMap.remove(token);
            }
        }, 10 * 60 * 1000);
        return token;
    }

    public Status handleConfirmation(String token) {
        if (registerUserMap.containsKey(token)) {
            RegisterUserDto regDto = registerUserMap.get(token);
            Status currentStatus = userValidator.checkUser(regDto);
            registerUserMap.remove(token);
            if (!currentStatus.isSuccess())
                return currentStatus;

            UserEntity member = saveUser(regDto, Role.USER);
            long pdaId = member.getPdaId();
            logger.info("Пользователь {} ({} {}) зарегистрирован.", member.getLogin(), member.getName(), member.getNickname());
            try {
                emailService.sendRegisterLetter(regDto, pdaId);
                return new Status(true, pdaId + " - Это ваш pdaId, мы вас зарегистрировали, спасибо!");
            } catch (Exception e) {
                logger.error("Handle confirmation", e);
                return new Status(true, "Не получилось отправить подтверждение на почту, но мы вас зарегистрировали, спасибо!");
            }
        } else return new Status(false, "Ссылка устарела или не существует");
    }

    public UserEntity saveUser(RegisterUserDto registerUserDto, Role role) {
        return userRepository.save(new UserEntity(registerUserDto, passwordEncoder, role));
    }

    public UUID getCurrentId() {
        return ((SecurityUser) SecurityContextHolder.getContext()
                .getAuthentication().getPrincipal()).getId();
    }

    @Override
    public UserEntity getUserById() {
        UserEntity userEntity = userRepository.findById(getCurrentId()).orElseThrow(() -> new RuntimeException("Пользователь не найден"));
        if (userEntity.getLastLoginAt().plusSeconds(300).isBefore(Instant.now())) {
            userEntity.setLastLoginAt(Instant.now());
            return userRepository.save(userEntity);
        } else return userEntity;
    }

    @Override
    public UserDto getUserDto() {
        return userMapper.dto(getUserById());
    }

    @Override
    public UserEntity getUserById(UUID objectId) {
        return userRepository.findById(objectId).orElseThrow();
    }

    @Override
    public AdminUserDto getUserForAdminById(UUID objectId) {
        return userMapper.adminDto(getUserById(objectId));
    }

    @Override
    public UserEntity getUserByEmail(String email) {
        return userRepository.findMemberByEmail(email)
                .orElseThrow(() -> new RuntimeException("Пользователя не существует"));
    }

    @Override
    public UserEntity getUserByLogin(String login) {
        return userRepository.findByLogin(login).orElseThrow(() -> new RuntimeException("Пользователя не существует"));
    }

    @Override
    @AdminAccess
    public AdminUserDto updateUser(UUID id, AdminEditUserDto adminEditUserDto) {
        UserEntity user = getUserById(id);
        user.setName(adminEditUserDto.getName());
        user.setNickname(adminEditUserDto.getNickname());
        user.setLogin(adminEditUserDto.getLogin());
        user.setEmail(adminEditUserDto.getEmail());
        user.setAvatar(adminEditUserDto.getAvatar());

        user.setRole(adminEditUserDto.getRole());
        user.setGang(adminEditUserDto.getGang());
        user.setChatBan(adminEditUserDto.getChatBan());
        logger.info("Пользователь {} обновлен модератором {}", userMapper.dto(user), getUserById().getLogin());

        return userMapper.adminDto(userRepository.save(user));
    }

    @Override
    public boolean setChatBan(UUID userId) {
        UserEntity user = getUserById(userId);
        user.setChatBan(!user.getChatBan());
        return userRepository.save(user).getChatBan();
    }


    @Override
    public Status editUser(RegisterUserDto user) {
        UserEntity userEntity = getUserById();
        userEntity.setName(user.getName());
        userEntity.setNickname(user.getNickname());
        //userEntity.setLogin(user.getLogin());
        //TODO добавить через модуль подтверждения почту
        //userEntity.setEmail(user.getEmail());
        userEntity.setAvatar(user.getAvatar());
        userEntity.setPassword(passwordEncoder.encode(user.getPassword()));
        userRepository.save(userEntity);

        String message = "Обновлен пользователь " + user.getLogin();
        return new Status(true, message);
    }

    @Override
    public void deleteUserById(UUID id) {
        userRepository.deleteById(id);
    }

    @Override
    public ByteArrayInputStream exportMailContacts() throws IOException {
        return exportUsers(userRepository.findAllByReceiveEmailsTrue());
    }

    @Override
    public boolean changeEmailSetting(UUID id) {
        UserEntity user = userRepository.findById(id).orElseThrow();
        user.setReceiveEmails(!user.getReceiveEmails());
        return userRepository.save(user)
                .getReceiveEmails();
    }

    public ByteArrayInputStream exportUsers(List<UserEntity> users) throws IOException {
        logger.info("{} exported {} contacts.", getUserById().getLogin(), users.size());
        XSSFWorkbook workbook = new XSSFWorkbook();
        Sheet sheet = workbook.createSheet("users");

        CellStyle headerCellStyle = workbook.createCellStyle();
        Row row = sheet.createRow(0);
        Cell cell = row.createCell(0);
        cell.setCellStyle(headerCellStyle);

        cell = row.createCell(1);
        cell.setCellStyle(headerCellStyle);


        Iterator<UserEntity> userIterator = users.listIterator();
        for (int i = 0; userIterator.hasNext(); i++) {
            UserEntity contactEntity = userIterator.next();
            Row header = sheet.createRow(i);

            Cell headerCell = header.createCell(0);
            headerCell.setCellStyle(headerCellStyle);
            headerCell.setCellValue(contactEntity.getEmail());

            headerCell = header.createCell(1);
            headerCell.setCellStyle(headerCellStyle);
            headerCell.setCellValue(contactEntity.getLogin());

            headerCell = header.createCell(2);
            headerCell.setCellStyle(headerCellStyle);
            headerCell.setCellValue(contactEntity.getName());

            headerCell = header.createCell(3);
            headerCell.setCellStyle(headerCellStyle);
            headerCell.setCellValue(contactEntity.getId().toString());
        }
        sheet.autoSizeColumn(0);
        sheet.autoSizeColumn(1);
        sheet.autoSizeColumn(2);

        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
        workbook.write(outputStream);
        workbook.close();
        return new ByteArrayInputStream(outputStream.toByteArray());
    }
}

File: pdanet/src/main/java/net/artux/pdanetwork/service/user/UserServiceImpl.kt
package net.artux.pdanetwork.service.user

import jakarta.mail.MessagingException
import mu.KLogging
import net.artux.pdanetwork.entity.mappers.UserMapper
import net.artux.pdanetwork.entity.security.SecurityUser
import net.artux.pdanetwork.entity.user.UserConfirmationEntity
import net.artux.pdanetwork.entity.user.UserEntity
import net.artux.pdanetwork.exception.UserNotFoundException
import net.artux.pdanetwork.models.Status
import net.artux.pdanetwork.models.user.dto.AdminEditUserDto
import net.artux.pdanetwork.models.user.dto.AdminUserDto
import net.artux.pdanetwork.models.user.dto.RegisterUserDto
import net.artux.pdanetwork.models.user.dto.UserDto
import net.artux.pdanetwork.models.user.enums.Role
import net.artux.pdanetwork.repository.user.UserConfirmationRepository
import net.artux.pdanetwork.repository.user.UserRepository
import net.artux.pdanetwork.service.email.EmailService
import net.artux.pdanetwork.service.util.ValuesService
import net.artux.pdanetwork.utils.RandomString
import net.artux.pdanetwork.utils.security.AdminAccess
import org.apache.commons.io.output.ByteArrayOutputStream
import org.apache.poi.ss.usermodel.CellStyle
import org.apache.poi.ss.usermodel.Sheet
import org.apache.poi.xssf.usermodel.XSSFWorkbook
import org.springframework.context.MessageSource
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.io.ByteArrayInputStream
import java.io.IOException
import java.time.Duration
import java.time.Instant
import java.util.Locale
import java.util.UUID

@Service
@Transactional
@Suppress("LongParameterList", "TooManyFunctions")
open class UserServiceImpl(
    private val userRepository: UserRepository,
    private val userConfirmationRepository: UserConfirmationRepository,
    private val emailService: EmailService,
    private val valuesService: ValuesService,
    private val userValidator: UserValidator,
    private val passwordEncoder: PasswordEncoder,
    private val userMapper: UserMapper,
    private val messageSource: MessageSource
) : UserService {

    private val randomString = RandomString()

    @Transactional
    override fun registerUser(registerUser: RegisterUserDto): Status {
        val status = userValidator.checkUser(registerUser)
        if (!status.isSuccess) {
            return status
        }
        return try {
            val createdUserEntity = saveUser(registerUser, INITIAL_ROLE)
            val token = generateConfirmationToken(createdUserEntity)

            emailService.sendConfirmLetter(registerUser, token)
            Status(
                true,
                """
                    Учетная запись зарегистрирована, но лучше ее подтвердить переходом по ссылке из письма,
                    которое было отравлено на ${registerUser.email}. Можете выполнить вход по этому email.
                """.trimIndent()
            )
        } catch (e: MessagingException) {
            logger.error("Registration", e)
            Status(false, "Не удалось отправить письмо на " + registerUser.email)
        }
    }

    private fun generateConfirmationToken(user: UserEntity): String {
        val token = randomString.nextString()
        logger.info(
            "Пользователь {} добавлен в базу с токеном подтверждения {}",
            user.email,
            token
        )

        logger.info("Ссылка подтверждения аккаунта: " + valuesService.address + "/confirmation/register?t=" + token)
        userConfirmationRepository.save(
            UserConfirmationEntity().apply {
                this.user = user
                this.token = token
            }
        )
        return token
    }

    @Transactional
    override fun handleConfirmation(token: String): Status {
        val confirmationEntity = userConfirmationRepository.findByToken(token)
        return if (confirmationEntity.isPresent) {
            val user = userRepository.findById(confirmationEntity.get().id).orElseThrow()
            user.isConfirmed = true
            userRepository.save(user)
            userConfirmationRepository.deleteById(user.id)

            logger.info("Пользователь {} ({} {}) подтвержден.", user.login, user.name, user.nickname)
            try {
                emailService.sendRegisterLetter(user)
                Status(true, "${user.pdaId} - Это ваш pdaId, мы вас зарегистрировали, спасибо!")
            } catch (e: MessagingException) {
                logger.error("Handle confirmation", e)
                Status(
                    true,
                    "Не получилось отправить подтверждение на почту, но мы вас зарегистрировали, спасибо!"
                )
            }
        } else {
            Status(false, "Ссылка устарела или не существует")
        }
    }

    private fun saveUser(registerUserDto: RegisterUserDto, role: Role): UserEntity {
        return userRepository.save(UserEntity(registerUserDto, passwordEncoder, role))
    }

    override fun getUsersByIds(ids: Collection<UUID>): List<UserEntity> {
        return userRepository.findAllById(ids)
    }

    override fun getCurrentId(): UUID {
        val securityUser = SecurityContextHolder.getContext().authentication.principal as SecurityUser
        return securityUser.id
    }

    override fun getCurrentUser(): UserEntity {
        val userEntity = userRepository.findById(getCurrentId()).orElseThrow {
            UserNotFoundException("Пользователь не найден")
        }

        val lastOnlineAt = userEntity.lastLoginAt.plus(UPDATE_ONLINE_DURATION)
        return if (lastOnlineAt.isBefore(Instant.now())) {
            updateLastLoginAt(userEntity)
        } else {
            userEntity
        }
    }

    @Transactional
    open fun updateLastLoginAt(userEntity: UserEntity): UserEntity {
        userEntity.lastLoginAt = Instant.now()
        return userRepository.save(userEntity)
    }

    override fun getUserDto(): UserDto {
        return userMapper.dto(getCurrentUser())
    }

    override fun getCurrentUser(objectId: UUID): UserEntity {
        return userRepository.findById(objectId).orElseThrow()
    }

    override fun getUserForAdminById(objectId: UUID): AdminUserDto {
        return userMapper.adminDto(getCurrentUser(objectId))
    }

    override fun getUserByEmail(email: String): UserEntity {
        return userRepository.findByEmail(email).orElseThrow { RuntimeException("Пользователя не существует") }
    }

    override fun getUserByLogin(login: String): UserEntity {
        return userRepository.findByLogin(login).orElseThrow { RuntimeException("Пользователя не существует") }
    }

    override fun getUsersByLogins(logins: Collection<String>): List<UserEntity> {
        return userRepository.findAllByLoginIn(logins)
    }

    @AdminAccess
    override fun updateUser(userId: UUID, adminEditUserDto: AdminEditUserDto): AdminUserDto {
        val user = getCurrentUser(userId)
        user.name = adminEditUserDto.name
        user.nickname = adminEditUserDto.nickname
        user.login = adminEditUserDto.login
        user.email = adminEditUserDto.email
        user.avatar = adminEditUserDto.avatar
        user.role = adminEditUserDto.role
        user.gang = adminEditUserDto.gang
        user.setChatBan(adminEditUserDto.chatBan)
        logger.info("Пользователь {} обновлен модератором {}", userMapper.dto(user), getCurrentUser().login)
        return userMapper.adminDto(userRepository.save(user))
    }

    override fun setChatBan(userId: UUID): Boolean {
        val user = getCurrentUser(userId)
        user.setChatBan(!user.chatBan)
        return userRepository.save(user).chatBan
    }

    override fun editUser(user: RegisterUserDto): Status {
        val userEntity = getCurrentUser()
        userEntity.nickname = user.nickname
        // userEntity.setLogin(user.getLogin());
        // TODO добавить через модуль подтверждения почту
        // userEntity.setEmail(user.getEmail());
        userEntity.avatar = user.avatar
        userEntity.password = passwordEncoder.encode(user.password)
        userRepository.save(userEntity)

        val message = messageSource.getMessage("user.updated", null, Locale.of("ru"))
        return Status(true, message)
    }

    override fun deleteUserById(id: UUID) {
        userRepository.deleteById(id)
    }

    @Throws(IOException::class)
    override fun exportMailContacts(): ByteArrayInputStream {
        return exportUsers(userRepository.findAllByReceiveEmailsTrue())
    }

    override fun changeEmailSetting(id: UUID): Boolean {
        val user = userRepository.findById(id).orElseThrow()
        user.receiveEmails = !user.receiveEmails
        return userRepository.save(user)
            .receiveEmails
    }

    @Throws(IOException::class)
    fun exportUsers(users: List<UserEntity>): ByteArrayInputStream {
        logger.info("{} exported {} contacts.", getCurrentUser().login, users.size)
        val workbook = XSSFWorkbook()
        val sheet: Sheet = workbook.createSheet("users")
        val headerCellStyle: CellStyle = workbook.createCellStyle()
        val row = sheet.createRow(0)
        var cell = row.createCell(0)
        cell.cellStyle = headerCellStyle
        cell = row.createCell(1)
        cell.cellStyle = headerCellStyle
        val userIterator: Iterator<UserEntity> = users.listIterator()
        var i = 0
        while (userIterator.hasNext()) {
            var cellCounter = 0
            val contactEntity = userIterator.next()
            val header = sheet.createRow(i)
            var headerCell = header.createCell(cellCounter++)
            headerCell.cellStyle = headerCellStyle
            headerCell.setCellValue(contactEntity.email)
            headerCell = header.createCell(cellCounter++)
            headerCell.cellStyle = headerCellStyle
            headerCell.setCellValue(contactEntity.login)
            headerCell = header.createCell(cellCounter++)
            headerCell.cellStyle = headerCellStyle
            headerCell.setCellValue(contactEntity.name)
            headerCell = header.createCell(cellCounter)
            headerCell.cellStyle = headerCellStyle
            headerCell.setCellValue(contactEntity.id.toString())
            i++
        }
        sheet.autoSizeColumn(0)
        sheet.autoSizeColumn(1)
        sheet.autoSizeColumn(2)
        val outputStream = ByteArrayOutputStream()
        workbook.write(outputStream)
        workbook.close()
        return ByteArrayInputStream(outputStream.toByteArray())
    }

    companion object : KLogging() {
        val INITIAL_ROLE = Role.USER
        val UPDATE_ONLINE_DURATION: Duration = Duration.ofMinutes(5)
    }
}
